File: ios/Ezber/Models/Progress.swift
import Foundation

/// Per-verse memorization state. Honest exposure, not a score.
enum VerseState: String, Codable, CaseIterable, Identifiable {
    case new
    case learning
    case review
    case strong

    var id: String { rawValue }

    var label: String {
        switch self {
        case .new: return "New"
        case .learning: return "Learning"
        case .review: return "Review"
        case .strong: return "Strong"
        }
    }

    var explanation: String {
        switch self {
        case .new: return "Not drilled yet."
        case .learning: return "Fewer than 10 repetitions so far."
        case .review: return "10–29 repetitions; worth reviewing."
        case .strong: return "30 or more repetitions."
        }
    }

    /// Provisional thresholds. The shared schema/domain work may define its own
    /// mapping; keep this the only place iOS derives a state from counts.
    static func derived(fromRepetitions repetitions: Int) -> VerseState {
        switch repetitions {
        case ..<1: return .new
        case 1..<10: return .learning
        case 10..<30: return .review
        default: return .strong
        }
    }
}

/// Accumulated exposure for one verse across all presets and sessions.
struct VerseProgress: Identifiable, Hashable, Codable {
    var verseID: VerseID
    var repetitionsCompleted: Int
    var exposureCount: Int
    var lastPlayedAt: Date?
    var state: VerseState
    var lastPresetID: UUID?

    var id: VerseID { verseID }

    init(
        verseID: VerseID,
        repetitionsCompleted: Int = 0,
        exposureCount: Int = 0,
        lastPlayedAt: Date? = nil,
        state: VerseState = .new,
        lastPresetID: UUID? = nil
    ) {
        self.verseID = verseID
        self.repetitionsCompleted = repetitionsCompleted
        self.exposureCount = exposureCount
        self.lastPlayedAt = lastPlayedAt
        self.state = state
        self.lastPresetID = lastPresetID
    }
}

File: android/app/src/main/java/app/ezber/android/models/Progress.kt
package app.ezber.android.models

/**
 * Per-(preset, verse) memorization state. The values are exactly the ones the
 * `progress.memorization_state` CHECK constraint allows in
 * schema/user_schema.sql: 'learning', 'review', 'memorized'.
 *
 * A verse with no `progress` row is shown as "New"; there is no stored new
 * state, which keeps "never drilled" honest.
 */
enum class MemorizationState(val label: String, val explanation: String) {
    LEARNING("Learning", "Fewer than 10 repetitions."),
    REVIEW("Review", "10–29 repetitions; still consolidating."),
    MEMORIZED("Memorized", "30 or more repetitions."),
    ;

    companion object {
        fun fromSchema(value: String?): MemorizationState =
            entries.firstOrNull { it.name.equals(value, ignoreCase = true) } ?: LEARNING

        /**
         * Provisional thresholds. Keep this the only place the app derives a
         * state from a repetition count; the shared schema may define its own
         * mapping later.
         */
        fun derived(repetitions: Int): MemorizationState = when {
            repetitions < 10 -> LEARNING
            repetitions < 30 -> REVIEW
            else -> MEMORIZED
        }
    }
}

/** One `progress` row: exposure for one preset's take on one verse. */
data class VerseProgress(
    val presetId: Long,
    val verseId: VerseId,
    val repetitionsDone: Int = 0,
    val state: MemorizationState = MemorizationState.LEARNING,
    val lastPlayedAt: String? = null,
    val nextReviewAt: String? = null,
)

/** One verse across every preset that has touched it. */
data class VerseSummary(
    val verseId: VerseId,
    val repetitions: Int,
    val exposureCount: Int,
    val lastPlayedAt: String?,
) {
    val state: MemorizationState get() = MemorizationState.derived(repetitions)
}

/**
 * Rolls per-(preset, verse) progress rows up to one entry per verse. The
 * progress screens show honest exposure across presets; `progress` itself
 * stays keyed by (preset_id, ayah) as schema/user_schema.sql requires.
 */
fun List<VerseProgress>.summarizeByVerse(): List<VerseSummary> =
    groupBy { it.verseId }
        .map { (verseId, rows) ->
            VerseSummary(
                verseId = verseId,
                repetitions = rows.sumOf { it.repetitionsDone },
                exposureCount = rows.size,
                lastPlayedAt = rows.mapNotNull { it.lastPlayedAt }.maxOrNull(),
            )
        }
        .sortedWith(compareBy({ it.verseId.surah }, { it.verseId.number }))
